README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
tycho2
======

Author: Eric Marty  
This R package provides a simple wrapper for the [Project
Tycho](https://www.tycho.pitt.edu/) 2.0 web API.

The development of this package was supported by the National Institute
of General Medical Sciences of the National Institutes of Health under
Award Number U01GM110744. The content is solely the responsibility of
the authors and does not necessarily reflect the official views of the
National Institutes of Health.

Project Tycho
-------------

Project Tycho is a repository for global health data in a standardized
format compliant with FAIR (Findable, Accessible, Interoperable, and
Reusable) guidelines.

Version 2.0 of the database currently contains:

-   weekly case counts for 78 notifiable conditions for 50 states and
    1284 cities between 1888 and 2014, reported by health agencies in
    the United States.
-   data for dengue-related conditions for 100 countries between 1955
    and 2010, obtained from the World Health Organization and national
    health agencies.

Project Tycho 2.0 datasets are represented in a standard format
registered with FAIRsharing (bsg-s000718) and include standard SNOMED-CT
codes for reported conditions, ISO 3166 codes for countries and first
administrative level subdivisions, and NCBI TaxonID numbers for
pathogens.

Precompiled datasets with DOI’s are also available for download directly
from the [Project Tycho](https://www.tycho.pitt.edu/).

See <https://www.tycho.pitt.edu/dataset/api/> for complete documentation
of the API.

Project Tycho database, website and API are by:  
[Wilbert van Panhuis](https://www.tycho.pitt.edu/people/person/49/)
(Principal Investogator)  
[Donald Burke](https://www.tycho.pitt.edu/people/person/66/) (Principal
Investogator)  
[Anne Cross](https://www.tycho.pitt.edu/people/person/50/) (Database
Programmer)  
Project Tycho is published under a [Creative Commons Attribution 4.0
International Public
License](http://creativecommons.org/licenses/by/4.0/).

tycho2 package
--------------

-   `apicall()` - Helper function to assemble arbitrary web api call
    URLs from a base URL and query string terms. Returns a URL as a
    string.
-   `tycho2()` - Wrapper for the Tycho 2.0 web API. Calls `apicall()`
    with Tycho’s default base URL and applies some automation to
    facilitate downloading large datasets with a single call. Returns a
    dataframe.


NEWS.md
# tycho2 0.3.0

* added argument to `apicall()` to specify reponse format
* `apicall()` now resolves incomplete urls and fixes missing and extra slashes

# tycho2 0.2.0

* update readme  
* added 'dataframe2zoo()' and 'tycho2zoo()'


# tycho2 0.1.0

* fixed classes of returned dataframe columns
* added documentation of returned dataframe columns
* added data object `weekdaynumbers`, a numeric vector storing numeric equivalents of named weekdays, 
including start and end days for the CDC Epidemiological week and ISO week.
* added function `round2wday()` to round dates to nearest weekday (specified as int, named weekday, 
start or end of CDC epiweek, or start or end of ISO week.)
* added `fixdates` parameter to `tycho2()`

# tycho2 0.0.1

* Initial release

# tycho2 0.0.0.9000

* initial deployment as a package



R/apicall.R
#' Assemble a web API call URL
#'
#' Assembles an arbitrary web API call URL from a base URL and query string terms. Returns a URL as
#' a string.
#'
#' A web API call consitst of a base URL with optional additional path, followed by a "?" separator
#' and query terms separated by the "&" separator.
#'
#' Query terms are in the form \code{[key][operator][value]}, most commonly \code{[key]=[value]}.
#'
#' \code{apicall()} uses the \code{baseurl} argument, if provided, as the base URL. \code{path} is
#' appended to \code{baseurl} before the "?" separator. Queries are built from a list of key-value
#' pairs passed to the \code{param} argument, and/or a character vector of query terms (conditions)
#' passed to the \code{queryterms} argument. An API Key, if required, can be set via the
#' \code{apikey} argument, or passed to \code{param} or \code{queryterms}. Any combination of
#' \code{queryterms}, \code{param} and \code{apikey} can be used.
#'
#' \code{apicall()} automatically replaces spaces with \code{\%20} in the final URL.
#'
#' @param baseurl string. The base URL to which an additional path and additional query terms may be
#'   appended. If baseurl does not begin with 'http://' or 'https://', then 'https://' is prepended.
#' @param path string (optional). An additional path to be appended to the base URL before the "?"
#'   separator and query terms.
#' @param params list (optional). A list of query terms in the form
#'   \code{list(key1=value1,key2=value2,...)}
#' @param queryterms character vector (optional). Vector of query terms passed as strings in the
#'   form \code{c("key1[operator]value1", "key2[operator]value2", ...)}. This is useful if the API
#'   allows operators other than \code{=}, for example \code{http://foo.foo/query?count>=20}.
#' @param apikey string. (optional). An API key. The API key can also be passed with \code{params}
#'   or \code{queryterms}.
#' @param format string. (optional). A reponse format given as an extension. If none is specified,
#' the API's default reponse format will be requested. Many API's use JSON as the default. Example
#' formats might include "csv", "xlm", etc. The format will be appended to the API call as a file
#' extension.
#'
#' @return string
#'
#' @examples
#' \dontrun{
#' apicall(baseurl="http://foo.com/", path="list")
#'
#' p <- list(author="Davy Jones",format="book")
#' apicall(baseurl="http://foo.com/", path="search", params=p)
#'
#' q <- c("author=Davy Jones","year>=1980")
#' apicall(baseurl="http://foo.com/", path="search", queryterms=q)
#'
#' apicall(baseurl="http://foo.com/search", params=p, queryterms="year>=1980", apikey="123abc")
#' }
#'
#' @export

apicall <- function(baseurl, path = "", params=NULL, queryterms=NULL, apikey=NULL, format=NULL) {
  b <- gsub('/$', '', baseurl[1]) # strip final slash
  if(!grepl('^http', b)){ b <- paste0("https://",b) } # resolve protocol
  c <- gsub('^/', '', path[1]) # strip initial slash
  c <- gsub('/$', '', c) # strip final slash
  u <- paste(b,c,sep="/") # concatenate

  q <- NULL
  if(!is.null(apikey)) {q <- c(q, paste0("apikey=",apikey))}
  if(!missing(params)) {q <- c(q, paste0(names(params),"=",params))}
  q <- c(q, queryterms)
  if (!is.null(q)) {
    qstring <- paste(q, collapse = "&")
    u <- paste(u,qstring,sep="?")
  }
  u <- gsub(" ","%20",u)
  if (!is.null(format)){
    f <- tolower(gsub('\\.', '', format))
    u <- paste(u,f,sep=".")
  }
  return(u)
}

### TODO Add support for REST API's with verbs.


R/round2wday.R
#' Round a date to the nearest specified weekday
#'
#' \code{round2wday} rounds a date to the nearest target weekday.
#'
#' Target weekday can be specified as a named weekday or its numerical equivalent,
#' from the following table:
#'
#' \tabular{rl}{
#'   \strong{Name} \tab \strong{Number}\cr
#'   Sunday \tab 1\cr
#'   Monday \tab 2\cr
#'   Tuesday \tab 3\cr
#'   Wednesday \tab 4\cr
#'   Thursday \tab 5\cr
#'   Friday \tab 6\cr
#'   Saturday \tab 7\cr
#'   cdcweek.start \tab 1\cr
#'   cdcweek.end \tab 7\cr
#'   isoweek.start \tab 2\cr
#'   isoweek.end \tab 1\cr
#'   }
#'
#' @param date Date, POSIXct, POSIXlt, or character string the the format "YYYY-MM-DD"
#' @param targetwday Target weekday
#' @return Date ("YYYY-MM-DD")
#' @export
round2wday <- function(date,targetwday) {
  if(is.numeric(targetwday)) {
    t.day <- as.integer(targetwday)
  }else{
    t.day <- as.integer(weekdaynumbers[targetwday])
  }
  if(is.na(t.day)) {
    stop(paste("targetweekday must be an integer 1 to 7, or one of:\n",
               paste(names(weekdaynumbers),collapse = ", ")))
  }
  i.date <- as.Date(date)
  i.day <- as.POSIXlt(i.date)$wday + 1L
  shift <- -(i.day - t.day + 4L) %% 7L - 3L
  return(i.date+shift)
}



R/dataframe2zoo.R
#' Extract ZOO Time Series of Cases from a Tycho 2.0 dataframe
#'
#' \code{extractzoo} and \code{tycho2zoo} extracts regular time series of class \code{zooreg} from a
#' dataframe or tibble.
#'
#' \code{extractzoo} aggregates the \code{data} by summing for each unique value of \code{index},
#' calculates the underlying time interval of the data, pads missing data with \code{NA}'s, and
#' returns a regular time series (numeric vector) of class \code{zooreg}. See
#' \code{\link[zoo]{zooreg}} for details.
#'
#' \code{tycho2zoo} is a convenience wrapper for \code{extractzoo} with default values
#' \code{index = "PeriodEndDate"} and \code{data = "CountValue"}.
#'
#' @param dataset Data Frame or Tibble with at least one variabel of class \code{Date},
#' \code{POSIXct} or \code{POSIXlt}.
#' @param index Character. Name of variable to use as index. Variable must be of class \code{Date},
#' \code{POSIXct} or \code{POSIXlt}
#' @param data Character. Name of variable to use as data.
#' @param FUN Name of function to use for aggregation. May be quoted or not. Default = \code{sum}
#'
#' @return A regular time series of class \code{zooreg}.
#'
#' @import dplyr
#' @import zoo
#' @importFrom padr pad
#' @importFrom rlang sym
#' @importFrom tibble as_tibble
#' @importFrom stats frequency
#'
#' @examples
#'
#' mydata <- data.frame(date = as.Date(c("2000/1/1",
#'                                           "2000/1/1",
#'                                           "2000/1/8",
#'                                           "2000/1/22",
#'                                           "2000/1/29",
#'                                           "2000/1/29")),
#'                          city = c("NY","BOS","NY","NY","BOS","NY"),
#'                          temp = c(2.0, 0.5, 5.5, 3.8, 7.1, 9.0))
#' mydata
#' meantemp <- dataframe2zoo(mydata, index="date", data="temp", FUN=mean)
#' meantemp
#'
#' @export

dataframe2zoo <- function(dataset, index, data, FUN="sum") {
  FUN <- match.fun(FUN)
  index <- rlang::sym(index)
  data <- rlang::sym(data)
  tbl <- tibble::as_tibble(dataset)
  tbl <- dplyr::group_by(tbl, !!index)
  tbl <- dplyr::summarise(tbl, newdata = FUN(!!data))
  tbl <- dplyr::arrange(tbl, !!index)
  tbl <- padr::pad(tbl)
  ts.zoo <- zoo::zoo(tbl$newdata, dplyr::pull(tbl, 1))
  attr(ts.zoo,"frequency") <- stats::frequency(ts.zoo)
  return(ts.zoo)
}

#' @rdname dataframe2zoo
#' @export

tycho2zoo <- function(dataset, index = "PeriodEndDate", data = "CountValue", FUN="sum") {
  dataframe2zoo(dataset = dataset, index = index, data = data)
}

## group_by_at(vars(-c(PlaceOfAcquisition,CountValue))) %>%  # more explicit
## group_by(ConditionName, PathogenName, Admin1Name, PeriodStartDate,PeriodEndDate) %>% # semi verbose


R/tycho2.R
#' Get data from Tycho 2.0 database
#'
#' Calls the Tycho 2.0 database using the Tycho 2.0 web API.
#'
#' \strong{Project Tycho, a repository for global health data}
#'
#' Project Tycho is a repository for global health data in a standardized format compliant with FAIR
#' (Findable, Accessible, Interoperable, and Reusable) guidelines.
#'
#' Version 2.0 of the database currently contains:
#'
#' \itemize{ \item Weekly case counts for 78 notifiable conditions for 50 states and 1284 cities
#' between 1888 and 2014, reported by health agencies in the United States. \item Data for
#' dengue-related conditions for 100 countries between 1955 and 2010, obtained from the World Health
#' Organization and national health agencies. }
#'
#' Project Tycho 2.0 datasets are represented in a standard format registered with FAIRsharing
#' (bsg-s000718) and include standard SNOMED-CT codes for reported conditions, ISO 3166 codes for
#' countries and first administrative level subdivisions, and NCBI TaxonID numbers for pathogens.
#'
#' Precompiled datasets with DOI's are also available for download directly from
#' \href{https://www.tycho.pitt.edu/}{Project Tycho}.
#'
#' See \url{https://www.tycho.pitt.edu/dataset/api/} for a complete documentation of the API.
#'
#' \strong{\code{tycho2()}}
#'
#' \code{tycho2} calls \code{\link{apicall}} with the base URL
#' "https://www.tycho.pitt.edu/api/". If \code{path} is the name of a data field in the Tycho 2.0
#' database, \code{tycho2} will return a dataframe of possible values for the field with additional
#' information. See \url{https://www.tycho.pitt.edu/dataset/api/} for more details. If \code{path}
#' is "query", \code{tycho2} will return a dataframe of case counts with associated variables for
#' the query terms specified. See \url{https://www.tycho.pitt.edu/dataset/api/} for more details.
#' Queries are built from a list of key-value pairs passed to the \code{param} argument, and/or a
#' character vector of query terms (conditions) passed to the \code{queryterms} argument. An account
#' with Project Tycho and an API Key is required to access the database.  The API Key can be
#' retrieved from your Tycho account. The API key can be set with the \code{apikey} argument, or
#' passed to \code{param} or \code{queryterms}. Any combination of \code{queryterms}, \code{param}
#' and \code{apikey} can be used.
#'
#' \code{tycho2()} automatically replaces spaces with \code{\%20} in the final URL.
#'
#' To pull large datasets, \code{tycho2()} repeatedly calls the API to retrieve partial datasets in
#' chunks of 5000 records until all the requested data has been received, then outputs a single
#' large dataframe. Therefore, the \code{limit} and \code{offset} querry parameters described in the
#' API do not need to be specified. \code{tycho2()} handles these parameters invisibly.
#'
#' To avoid errors, date ranges should be specified in YYYY-MM-DD format using
#' \code{PeriodStartDate} and \code{PeriodEndDate} query parameters with the \code{>=} and \code{<=}
#' operators. The use of \code{>=} and \code{<=} requires passing dates using the "queryterms"
#' argument.
#'
#' Although the Tycho 2.0 database can be querried directly by passing a manually assembled API call
#' URL to \code{read.csv}, as below...
#'
#' \code{read.csv('https://www.tycho.pitt.edu/api/query?CountryISO=US&ConditionName=Gonorrhea&apikey=YOURAPIKEY')}
#'
#' ...use of \code{tycho2} allows querries to be assembled more flexibly and programmatically.
#'
#' Accessing the Project Tycho API using \code{tycho2} requires an API key, which can be retrieved
#' from your Project Tycho account. You must have a Project Tycho account to receive an API key.
#'
#' The Project Tycho 2.0 database and API are by
#' \href{https://www.tycho.pitt.edu/people/person/49/}{Wilbert van Panhuis} (Principal
#' Investogator), \href{https://www.tycho.pitt.edu/people/person/66/}{Donald Burke} (Principal
#' Investogator), \href{https://www.tycho.pitt.edu/people/person/50/}{Anne Cross} (Database
#' Programmer). Project Tycho is published under a
#' \href{http://creativecommons.org/licenses/by/4.0/}{Creative Commons Attribution 4.0 International
#' Public License}.
#'
#' @param path string (optional). Must be either "query" to perform data queries, or one of the
#'   tycho 2.0 database fields to retrieve variable listings.
#' @param params list (optional). A list of query terms in the form
#'   \code{list(var1=value1,var2=value2,...)}
#' @param queryterms character vector (optional). Vector of query terms passed as strings in the
#'   form \code{c("var1[operator]value1", "var2[operator]value2", ...)}. Dates must be passed this
#'   way using the \code{>=} and \code{<=} operators (i.e.
#'   \code{queryterms=c("PeriodStartDate>=2000-01-01")})
#' @param apikey string. (required).   Your Project Tycho API key. This can also be passed with
#'   \code{params} or \code{queryterms}.
#' @param baseurl string. Defaults to "https://www.tycho.pitt.edu/api/".
#' @param fixdates "cdc", "iso", or NULL.  If \code{fixdates="cdc"}, PeriodStartDate and PeriodEndDate
#' are rounded to nearest CDC epidemiological week ("epiweek") start and end days
#' (Sunday - Saturday), respectively. If \code{fixdates="cdc"}, PeriodStartDate and PeriodEndDate are
#' rounded to nearest ISO week start and end dates (Monday - Sunday), respectively. CDC epiweeks are
#' used for US data reporting. Elsewhere, epiweeks are synonymous with ISO weeks. Rounding is done
#' with \code{\link{round2wday}}. This param may be necessary because some entries in the Tycho 2.0
#' database have incorrect dates that may be off by one day from the actual epiweek start or end
#' dates. default=NULL.
#' @param start Date, POSIXct, POSIXlt, or character string in "YYYY-MM-DD" format. The start date.
#' If present, overrides "PeriodStartDate" passed to \code{queryterms}. Default = NULL
#' @param end Date, POSIXct, POSIXlt, or character string in "YYYY-MM-DD" format. The end date.
#' If present, overrides "PeriodEndDate" passed to \code{queryterms}. Default = NULL
#'
#' @return dataframe with the following possible columns:
#'   \item{$ConditionName}{factor. Name of reported condition as listed in \href{https://doi.org/10.25504/FAIRsharing.d88s6e}{SNOMED-CT}}
#'   \item{$ConditionSNOMED}{factor. \href{https://doi.org/10.25504/FAIRsharing.d88s6e}{SNOMED-CT} code for reported condition}
#'   \item{$PathogenName}{factor. \href{https://doi.org/10.25504/FAIRsharing.fj07xj}{NCBI Taxonomy} organism name for pathogen causing reported condition}
#'   \item{$PathogenTaxonID}{factor. \href{https://doi.org/10.25504/FAIRsharing.fj07xj}{NCBI Taxonomy} identifier for pathogen causing reported condition}
#'   \item{$Fatalities}{logical. Counts of reported condition ($CountValue) represent fatalities}
#'   \item{$CountryName}{factor. \href{https://www.iso.org/obp/ui/#search/code/}{ISO 3166} English Short Name of country}
#'   \item{$CountryCode}{factor. \href{https://www.iso.org/obp/ui/#search/code/}{ISO 3166} 2-letter code for country}
#'   \item{$Admin1Name}{factor. \href{https://www.iso.org/standard/63546.html}{ISO 3166-2} Name of first administrative subdivision (such as US state)}
#'   \item{$Admin1ISO}{factor. \href{https://www.iso.org/standard/63546.html}{ISO 3166-2} code for first administrative subdivision}
#'   \item{$Admin2Name}{factor. \href{http://www.geonames.org/}{Geonames} Placename of second order administrative division}
#'   \item{$CityName}{factor. \href{http://www.geonames.org/}{Geonames} Name of populated place}
#'   \item{$PeriodStartDate}{Date, format: YYYY-MM-DD. Start date of time interval for which a count was reported}
#'   \item{$PeriodEndDate}{Date, format: YYYY-MM-DD. End date of time interval for which a count was reported}
#'   \item{$PartOfCumulativeCountSeries}{logical. Count is part of a series of cumulative counts
#'   (instead of being part of a series of fixed time interval counts)}
#'   \item{$AgeRange}{Ordered factor. Age range in years for which a count was reported e.g. "0-18". Max age = 130}
#'   \item{$Subpopulation}{factor. "Civilian", "Military", or "None specified"}
#'   \item{$PlaceOfAcquisition}{factor. "Domestic", "Abroad", or NA}
#'   \item{$DiagnosisCertainty}{factor. \href{https://doi.org/10.25504/FAIRsharing.d88s6e}{SNOMED-CT}
#'   Qualifier for certainty of diagnosis for a count condition: "Definite", "Equivocal", "Possible diagnosis","Probable diagnosis", or NA}
#'   \item{$SourceName}{factor. Name of the source (system, database, institution) from which counts
#'   were obtained by the Project Tycho team}
#'   \item{$CountValue}{integer. The count value.}
#'   Variables described in detail here:
#'   \url{https://www.tycho.pitt.edu/dataformat/ProjectTychoCustomCompiledDataFormat.pdf}
#
#'
#' @examples
#' \dontrun{
#' # Note: retrive your API key from your Project Tycho account
#'
#' # List of conditions showing "ConditionName", "ConditionSNOMED"
#'
#' TYCHOKEY <- 'some1long2alphanumeric3string'
#' conditions <- tycho2("condition", apikey = TYCHOKEY)
#'
#' # All cases of scarlet fever in California
#'
#' params <- list(ConditionName = "Scarlet fever", Admin1ISO = "US-CA")
#' Scarlet <- tycho2("query", params = params, apikey = TYCHOKEY)
#'
#' # All measles cases in California from 2000 to 2010
#'
#' queryterms <- c(
#'   "ConditionName=Measles",
#'   "Admin1ISO=US-CA",
#'   "PeriodStartDate>=2000-01-01",
#'   "PeriodEndDate<=2010-01-01"
#'   )
#' Measles_CA_2000_2010 <- tycho2("query", queryterms=queryterms, apikey=TYCHOKEY)
#' }
#'
#' @export
#' @importFrom utils read.csv
#'
tycho2 <- function(path="", params=NULL, queryterms=NULL, apikey=NULL,
                   baseurl="https://www.tycho.pitt.edu/api/",
                   fixdates=NULL, start=NULL, end=NULL){
  p <- params
  p$offset <- 0
  p$limit <- 5000
  q <- queryterms[grep("offset",queryterms,invert = TRUE)]
  q <- queryterms[grep("limit",queryterms,invert = TRUE)]

  if(!is.null(start)){
    q <- c(
      q[grep("PeriodStartDate",q,invert = TRUE)],
      paste0("PeriodStartDate>=",as.Date(start))
      )
  }
  if(!is.null(end)){
    q <- c(
      q[grep("PeriodEndDate",q,invert = TRUE)],
      paste0("PeriodEndDate<=",as.Date(end))
    )
  }

  out <- utils::read.csv(apicall(baseurl=baseurl,path=path,params=p,queryterms=q,apikey=apikey))
  more <- nrow(out)>=5000
  while (more == TRUE) {
    p$offset <- p$offset+5000
    df <- read.csv(apicall(baseurl=baseurl, path, params=p, queryterms=q, apikey=apikey))
    more <- nrow(df)>0
    if (more) {
      out <- rbind(out,df)
    }
  }

  if(!is.null(fixdates)) {
    if (fixdates %in% c("cdc","iso")) {
      week.start <- weekdaynumbers[paste0(fixdates,"week.start")]
      week.end <- weekdaynumbers[paste0(fixdates,"week.end")]
    }else{
      warning('"fixdates" must be "cdc", "iso" or NULL. Ignoring "fixdates".')
    }
  }


  # classes
  vars <- colnames(out)

  if("ConditionName" %in% vars) {
    out$ConditionName <- as.factor(out$ConditionName)
  }
  if("ConditionSNOMED" %in% vars) {
    out$ConditionSNOMED <- as.factor(out$ConditionSNOMED)
  }
  if("PathogenName" %in% vars) {
    out$PathogenName <- as.factor(out$PathogenName)
  }
  if("PathogenTaxonID" %in% vars) {
    out$PathogenTaxonID <- as.factor(out$PathogenTaxonID)
  }
  if("Fatalities" %in% vars) {
    out$Fatalities <- as.logical(out$Fatalities)
  }
  if("CountryName" %in% vars) {
    out$CountryName <- as.factor(out$CountryName)
  }
  if("CountryCode" %in% vars) {
    out$CountryCode <- as.factor(out$CountryCode)
  }
  if("Admin1Name" %in% vars) {
    out$Admin1Name <- as.factor(out$Admin1Name)
  }
  if("Admin1ISO" %in% vars) {
    out$Admin1ISO <- as.factor(out$Admin1ISO)
  }
  if("Admin2Name" %in% vars) {
    out$Admin2Name <- as.factor(out$Admin2Name)
  }
  if("CityName" %in% vars) {
    out$CityName <- as.factor(out$CityName)
  }
  if("PeriodStartDate" %in% vars) {
    out$PeriodStartDate <- as.Date(out$PeriodStartDate,format = "%Y-%m-%d")
    if(!is.null(fixdates)) {
      out$PeriodStartDate <- round2wday(out$PeriodStartDate,week.start)
    }
  }
  if("PeriodEndDate" %in% vars) {
    out$PeriodEndDate <- as.Date(out$PeriodEndDate,format = "%Y-%m-%d")
    if(!is.null(fixdates)) {
      out$PeriodEndDate <- round2wday(out$PeriodEndDate,week.end)
    }
  }
  if("PartOfCumulativeCountSeries" %in% vars) {
    out$PartOfCumulativeCountSeries <- as.logical(out$PartOfCumulativeCountSeries)
  }
  if("AgeRange" %in% vars) {
    out$AgeRange <- as.ordered(out$AgeRange)
  }
  if("Subpopulation" %in% vars) {
    out$Subpopulation <- factor(out$Subpopulation,
                                levels = c("Civilian", "Military","None specified"))
  }
  if("PlaceOfAcquisition" %in% vars) {
    out$PlaceOfAcquisition <- factor(out$PlaceOfAcquisition,
                                     levels = c("Domestic", "Abroad"))
  }
  if("DiagnosisCertainty" %in% vars) {
    out$DiagnosisCertainty <- as.factor(out$DiagnosisCertainty)
  }
  if("SourceName" %in% vars) {
    out$SourceName <- as.factor(out$SourceName)
  }
  if("CountValue" %in% vars) {
    out$CountValue <- as.integer(out$CountValue)
  }

  return(out)
}




